app/src/parser_impl_masp.c

```diff
@@ -332,6 +336,14 @@ static parser_error_t readSaplingOutputDescriptionInfo(parser_context_t *ctx, ma
     }
 
     CHECK_ERROR(readUint32(ctx, &builder->n_outputs))
+#if defined(LEDGER_SPECIFIC) && !defined(APP_TESTING)
+    if (G_io_apdu_buffer[OFFSET_INS] == INS_SIGN_MASP_SPENDS) {
+        uint32_t rnd_outputs = (uint32_t)transaction_get_n_outputs();
+        if (rnd_outputs < builder->n_outputs) {
+            return parser_invalid_number_of_outputs;
+        }
+    }
+#endif
```

Comment:
This seems like something we can extract to a function, as it is repeated in many places. Not sure if stack usage will allow us to do it or not.